eth/polls/eth_subscribe.go

```diff
@@ -26,60 +27,79 @@ type EthPoll interface {
 
 type EthSubscriptions struct {
 	polls map[string]EthPoll
-
-	lastPoll    map[string]uint64
-	timestamps  map[uint64][]string
-	lastPrune   uint64
-	evmAuxStore *evmaux.EvmAuxStore
-	blockStore  store.BlockStore
+	lastPoll        map[string]uint64
+	timestamps      map[uint64][]string
+	mutex      *sync.RWMutex
+	lastPrune       uint64
+	evmAuxStore     *evmaux.EvmAuxStore
+	blockStore      store.BlockStore
 }
 
 func NewEthSubscriptions(evmAuxStore *evmaux.EvmAuxStore, blockStore store.BlockStore) *EthSubscriptions {
 	p := &EthSubscriptions{
-		polls:       make(map[string]EthPoll),
-		lastPoll:    make(map[string]uint64),
-		timestamps:  make(map[uint64][]string),
-		evmAuxStore: evmAuxStore,
-		blockStore:  blockStore,
+		polls:                  make(map[string]EthPoll),
+		lastPoll:               make(map[string]uint64),
+		timestamps:             make(map[uint64][]string),
+		mutex:             new(sync.RWMutex),
+
+		evmAuxStore:            evmAuxStore,
+		blockStore:             blockStore,
 	}
 	return p
 }
 
-func (s EthSubscriptions) Add(poll EthPoll, height uint64) string {
+func (s *EthSubscriptions) Add(poll EthPoll, height uint64) string {
 	id := utils.GetId()
-	s.polls[id] = poll
 
+	s.mutex.Lock()
+	s.polls[id] = poll
 	s.lastPoll[id] = height
 	s.timestamps[height] = append(s.timestamps[height], id)
+	s.mutex.Unlock()
+
 	s.pruneSubs(height)
 
 	return id
 }
 
-func (s EthSubscriptions) pruneSubs(height uint64) {
+func (s *EthSubscriptions) pruneSubs(height uint64) {
 	if height > BlockTimeout {
 		for h := s.lastPrune; h < height-BlockTimeout; h++ {
+			idsToRemove := []string{}
+
+			s.mutex.RLock()
```

Comment:
Just do this up top:
```golang
s.mutex.Lock()
defer s.mutex.Unlock()
```
This fine grained locking and unlocking is overly complicated.

Reply:
Something is need to avoid deadlock with 
```go
				s.Remove(id)
```.
`s.Remove(id)` needs to lock `polls` and `lastpoll`, its exported so we need the locks here.
This could also be done by splitting into three mutexes as the top-level only needs to lock `timestamps` while `Remove` only needs to lock `polls` and `lastPoll`